src/jrd/recsrc/IndexTableScan.cpp

```diff
@@ -492,33 +547,32 @@ UCHAR* IndexTableScan::getPosition(thread_db* tdbb, Impure* impure, win* window)
 
 UCHAR* IndexTableScan::openStream(thread_db* tdbb, Impure* impure, win* window) const
 {
-	// initialize for a retrieval
-	if (!setupBitmaps(tdbb, impure))
-		return NULL;
+	temporary_key* lower = impure->irsb_nav_current_lower;
+	temporary_key* upper = impure->irsb_nav_current_upper;
+	const bool firstKeys = lower == impure->irsb_nav_lower;
 
 	setPage(tdbb, impure, NULL);
 	impure->irsb_nav_length = 0;
 
 	// Find the starting leaf page
 	const IndexRetrieval* const retrieval = m_index->retrieval;
 	index_desc* const idx = (index_desc*) ((SCHAR*) impure + m_offset);
-	temporary_key lower, upper;
-	Ods::btree_page* page = BTR_find_page(tdbb, retrieval, window, idx, &lower, &upper);
+	Ods::btree_page* page = BTR_find_page(tdbb, retrieval, window, idx, lower, upper, firstKeys);
 	setPage(tdbb, impure, window);
 
 	// find the upper limit for the search
 	temporary_key* limit_ptr = NULL;
 	if (retrieval->irb_upper_count)
 	{
-		// If upper key length is greater than declared key length, we need 
-		// one "excess" byte for correct comparison. Without it there could 
+		// If upper key length is greater than declared key length, we need
+		// one "excess" byte for correct comparison. Without it there could
 		// be false equality hits.
-		impure->irsb_nav_upper_length = MIN(m_length + 1, upper.key_length);
-		memcpy(impure->irsb_nav_data + m_length, upper.key_data, impure->irsb_nav_upper_length);
+		impure->irsb_nav_upper_length = MIN(m_length + 1, upper->key_length);
+		memcpy(impure->irsb_nav_data + m_length, upper->key_data, impure->irsb_nav_upper_length);
 	}
 
 	if (retrieval->irb_lower_count)
-		limit_ptr = &lower;
+		limit_ptr = lower;
 
```

Comment:
Similarly to btr.cpp, jumping through keys via BTR_find_leaf() would make the patch against IndexScanNode.cpp 10x smaller and supposedly not slower. Something like this in `getRecord()`:

```c++
if (retrieval->irb_upper_count &&
  compareKeys(idx, key.key_data, key.key_length, &upper, flags) > 0)
// here the new code starts
{
  if (! get new lower/upper key)
    break;

  key = upper = new key;

  while (!(nextPointer = BTR_find_leaf(page, &upper, impure->irsb_nav_data, NULL,
            (idx->idx_flags & idx_descending), (retrieval->irb_generic & (irb_starting | irb_partial)))))
  {
    page = (Ods::btree_page*) CCH_HANDOFF(tdbb, &window, page->btr_sibling, LCK_read, pag_index);
  }

  continue;
}
```

Reply:
I suppose this and the find_node_start_point comments should not be changed then due to probable distance of keys, correct?

Reply:
Yes, my suggestion is not going to suit you well.